File: src/main/java/com/medicalsystem/importer/RowImporter.java
package com.medicalsystem.importer;

import com.medicalsystem.model.FormType;
import com.medicalsystem.util.CellUtils;
import lombok.AllArgsConstructor;
import lombok.extern.java.Log;
import org.apache.poi.ss.usermodel.Cell;
import org.apache.poi.ss.usermodel.Row;
import org.springframework.beans.factory.annotation.Autowired;
import org.springframework.stereotype.Component;

import java.util.Iterator;

@Component
@AllArgsConstructor(onConstructor = @__(@Autowired))
@Log
public class RowImporter {

    private final CellImporter cellImporter;

    /**
     * Processes a single row and imports its content to the database
     *
     * @param row an excel row
     */
    public void importRow(Row row, FormType formType, int maxNumberOfCells) {
        Iterator<Cell> iterator = row.iterator();

        // Retrieve patient ID from the first cell
        String patientIdStr = CellUtils.getStringValue(iterator.next());
        int patientId;
        try {
            patientId = Integer.parseInt(patientIdStr);
        } catch (NumberFormatException e) {
            log.warning(String.format("Patient ID is not an integer in row: '%d', value: '%s' - PATIENT NOT SAVED", row.getRowNum(), patientIdStr));
            return;
        }

        // Process the rest of the cells handling empty cells
        for (int i = 1; i < maxNumberOfCells; i++) {
            Cell cell;
            if (row.getCell(i) == null)
                cell = row.createCell(i);
            else
                cell = row.getCell(i);
            cellImporter.importCell(cell, patientId, formType);
        }

    }

}

File: src/main/kotlin/com/medicalsystem/excel/importer/RowImporter.kt
package com.medicalsystem.excel.importer

import com.medicalsystem.domain.Patient
import com.medicalsystem.excel.getAsString
import com.medicalsystem.excel.importer.result.ImportError
import com.medicalsystem.service.PatientService
import org.apache.poi.ss.usermodel.Cell
import org.apache.poi.ss.usermodel.Row
import org.springframework.stereotype.Component

@Component
class RowImporter(private val patientService: PatientService, private val cellImporter: CellImporter) {

    fun importRow(row: Row, formSheet: FormSheet): List<ImportError> {
        if (row.physicalNumberOfCells == 0) {
            return listOf(ImportError.noDataCellsInRow(formSheet, row))
        }

        // Extract patient ID
        val patientId = row.getPatientId()
        if (patientId == "") {
            return listOf(ImportError.patientIdEmpty(formSheet, row))
        }

        // Check if patient exists
        if (patientService.exists(patientId)) {
            return listOf(ImportError.patientExists(formSheet, row, patientId))
        }

        // Create and save Patient object
        val patient = patientService.save(Patient(patientId, formSheet.form))

        // Process the rest of the cells
        return (1 until formSheet.maxNumberOfCells)
                // Get cell or create empty cell if one does not exist at given index
                .map { row.getCell(it) ?: row.createCell(it) }
                // Import each cell and aggregate results
                .flatMap { cellImporter.importCell(it, formSheet, patient) }
    }

    private fun Row.getPatientId(): String {
        val cellWithId: Cell = this.iterator().next()
        return cellWithId.getAsString()
    }
}